src/nn/var_store.rs

```diff
@@ -59,6 +60,55 @@ impl VarStore {
         VarStore { variables_: Arc::new(Mutex::new(variables)), device }
     }
 
+    pub fn merge(var_stores: Vec<(VarStore, Option<&str>)>) -> Result<VarStore, TchError> {
+        let mut new_var_store = VarStore::new(Device::Cpu);
+
+        if var_stores.is_empty() {
+            Ok(new_var_store)
+        } else {
+            let mut new_variables =
+                Variables { named_variables: HashMap::new(), trainable_variables: Vec::new() };
+            let device = var_stores[0].0.device();
+
+            for (var_store, prefix) in var_stores {
+                if var_store.device() != device {
+                    return Err(TchError::Torch(format!(
+                        "All VarStores must be on the same device, got {:?} and {:?}",
+                        device,
+                        var_store.device()
+                    )));
+                }
+                for (var_name, var) in var_store.variables() {
+                    let new_var_name = format!("{}{}", prefix.unwrap_or(""), var_name);
+                    match new_variables.named_variables.entry(new_var_name) {
+                        Occupied(v) => {
+                            return Err(TchError::Torch(format!(
+                                "Duplicate variable name found: {}. Provide a unique prefix to allow merge operation",
+                                v.key(),
+                            )));
+                        }
+                        Vacant(v) => {
+                            v.insert(var);
+                        }
+                    }
+                }
+                for trainable_var in Arc::try_unwrap(var_store.variables_)
```

Comment:
What guarantees do we have that this would not panic? Maybe cloning the variables would be safer if this makes sense?

Reply:
Thank you this is a good point. I am unsure about the implication of cloning variables (do the new `VarStore` variables and tensor still correspond to the variables that may be loaded into a model in the input varstores?). I have instead updated it and not assume no other thread has a copy of the `Arc`, but rather lock it and drain the underlying variables vector. This means the older `VarStore` will be invalid (empty variables) but this was the intention of this method in any case.

Reply:
Right, as it's passing the argument `VarStore` as values and so consuming them, this seem reasonable.